### DISCUSSION TITLE
Excel worksheet access example

### DISCUSSION THREAD

**Discussion by jacek-kurlit** [sreeise/graph-rs-sdk#446] [Q&A]
Hi,

I need to access data from excel. I have found API description to do so https://learn.microsoft.com/en-us/graph/api/worksheet-usedrange?view=graph-rest-1.0&tabs=http
but I cannot find any example how to do this with this crate.
Did I missed something or there is no such feature yet?

**Comment by sreeise** [comment]
Hey @jacek-kurlit 

Workbooks is not included, or at least not yet. Thanks for reminding me about this though as I need to create a ticket for it.

I have wanted to add it already but its a large undertaking.

Basically the reason workbooks hasn't been added is because its a very very large set of APIs and the OpenAPI metadata document has a lot of issues or errors that would need to be manually reviewed and accounted for when generating the code. 

What will probably need to be done is to just do a limited subset of the APIs to start out with. I'll create the ticket and while I can't promise it will be done in a few weeks, APIs that need to be added are considered high priority.

Can you tell if are you just looking to use the Worksheets APIs or are there other parts of the Workbooks APIs your using?

Created #447

**Reply by jacek-kurlit** [reply]
Hi,

Thank you for the answer.
My use case is to call used range or range API to access data from excel sheet. No modifications whatsoever, read only.

If you need any help with that I may try to create PR to cover some cases.

Best Regards

**Reply by sreeise** [reply]
Thanks for letting me know. And feel free to create a PR. The contributions are welcome.

**Reply by jacek-kurlit** [reply]
I would love to contribute if I will find time!
Can you guide me where to start?

**Reply by sreeise** [reply]
Yeah for sure, so the wiki has a guide on how to generate APIs. It looks like all of the workbook apis are actually a sub-api of the drive resource. Which is good because its easier to generate the API when its a sub-resource.

Heres the wiki: [Codegen - Generating API's](https://github.com/sreeise/graph-rs-sdk/wiki/Codegen-%E2%80%90-Generating-API's)

I went ahead and made a few changes on a branch just to get you started: https://github.com/sreeise/graph-rs-sdk/compare/workbook

To get started on that branch you just need to create a rust file in the examples directory in order to run the codegen. Typically I call it test.rs and that filename has been added to the gitignore so you can't accidently push it up GitHub.

And once on that branch and in the test.rs file, run the following:

```rust
use graph_codegen::macros::OpenApiParser;
use graph_codegen::openapi::OpenApi;
use graph_codegen::settings::get_write_configuration;
use graph_core::resource::ResourceIdentity;

fn main() {
    OpenApi::write_all(
        vec![
            get_write_configuration(ResourceIdentity::Workbook),
            get_write_configuration(ResourceIdentity::WorkbookTables),
            get_write_configuration(ResourceIdentity::Worksheets),
            get_write_configuration(ResourceIdentity::WorksheetTableColumns),
        ],
    );
}
```

go to the drives folder located at src/drives folder in the repo, located here online https://github.com/sreeise/graph-rs-sdk/tree/master/src/drives and you will see four new folders. Add the names of the folders to src/drives/mod.rs 

Once youve added the module names to mod.rs you will probably want to run cargo fmt just to format the output so its more readable. 

Since workbook is a sub-resource of drive it was added as a child of the drive folder in the codegen settings. Which means the modules get generated in the drives folder. If you would like to regenerate the whole drives module, which will add the workbook modules to the mod.rs automatically you can just as easily run:

```rust
use graph_codegen::macros::OpenApiParser;
use graph_codegen::openapi::OpenApi;
use graph_codegen::settings::get_write_configuration;
use graph_core::resource::ResourceIdentity;

fn main() {
    OpenApi::write(get_write_configuration(ResourceIdentity::Drives));
}
```


And from there just follow the guide. It will walk through what I did in the branch in the very first part as well.

I assumed here it would probably be best to limit the scope of what APIs are generated for workbook and so I stuck to worksheets and ignored the rest as you will see in the code. But feel free to take on as much as you want. 

Let me know if you have any questions.

**Reply by jacek-kurlit** [reply]
Wow thank you for the answer! I will try to add code in free time.

**Reply by jacek-kurlit** [reply]
I have created initial PR https://github.com/sreeise/graph-rs-sdk/pull/448
I'm not sure what next or even how to test this?

**Reply by sreeise** [reply]
I'll reploy to the PR here in the next couple days. Thanks!
